selene_sdk/samplers/multi_sampler.py

```diff
@@ -0,0 +1,324 @@
+"""
+This module provides the `MultiSampler` class, which uses a
+FileSampler for each mode of sampling (train, test, validation).
+The MultiSampler is therefore a subclass of Sampler.
+"""
+import numpy as np
+from torch.utils.data import DataLoader
+
+from .sampler import Sampler
+
+def MultiFileSampler(*args, **kwargs):
+    from warnings import warn
+    warn("MultiFileSampler is deprecated and will be removed from future "
+    "versions of Selene. Please use MultiSampler instead.")
+    return MultiSampler(*args, **kwargs)
+
+class MultiSampler(Sampler):
+    """
+    This sampler draws samples from individual file samplers or data loaders
+    that corresponds to training, validation, and testing (optional) modes.
+    MultiSampler calls on the correct file sampler or data loader to draw
+    samples for a given mode. Example file samplers are under
+    `selene_sdk.samplers.file_samplers` and example data loaders are under
+    `selene_sdk.samplers.dataloaders`.
+
+    MultiSampler can use either file samplers or data loaders for
+    different modes. Using data loaders for some modes while using file samplers
+    for other modes are also allowed. The file samplers parse data files
+    (e.g. bed, mat, or hdf5). The data loaders provide multi-worker iterators
+    that draw samples from online samplers (i.e. on-the-fly sampling). As data
+    loaders support parallel sampling, they are generally recommended for
+    sampling speed.
+
+    Parameters
+    ----------
+    train_sampler : selene_sdk.samplers.file_samplers.FileSampler or \
+                    selene_sdk.samplers.dataloader.DataLoader
+        Load your training data as a `FileSampler` or `DataLoader`
+    validate_sampler : FileSampler or DataLoader
+        The validation dataset file sampler or data loader.
+    features : list(str)
+        The list of features the model should predict
+    test_sampler : None or FileSampler or DataLoader, optional
+        Default is None. The test file sampler is optional.
+    mode : str, optional
+        Default is "train". Must be one of `{train, validate, test}`. The
+        starting mode in which to run the sampler.
+    save_datasets : list(str) or None, optional
```

Comment:
I'm still thinking about this haha - for MultiSampler actually the `save_dataset` functionality could be called in the other train/valid/test sampler classes right? So maybe it's not useless as long as those classes have the functionality in them... I'm going to check and think about this more

Reply:
Yeah if `MultiSampler` uses `FileSampler` it should be straightforward to implement `save_dataset` by passing it to the FileSampler. For DataLoader because there are multiple instances of sampler due to multiprocessing to support it we will probably have to ask each sampler to return the start/end etc information and return maybe as a (seq, target, info) tuple and handle it in DataLoader.  it's a bit more complicated as we may need a custom collate_fn (https://pytorch.org/docs/stable/data.html#dataloader-collate-fn; used by dataloader to combine multiple returned values into one)

Reply:
oh right, thanks I forgot about the DataLoader issue. ok we can leave this for now and maybe I'll continue thinking about it before we push this branch hahah